### DISCUSSION TITLE
Quick way to find a point inside a manifold object

### DISCUSSION THREAD

**Discussion by pyenthu** [elalish/manifold#885] [General]
Is there a fast way to find whether a point is inside this manifold object based on the structure you have?
My objective is to perform some tasks with vertices of another mesh that intersects a manifold object without actually doing a Boolean operation. 
What I am looking for is to iterate over its vertices to see if they lie in the manifold. It has to be fast and near real time.

**Comment by pca006132** [comment]
We do have a bvh-based collider, but it is not exposed to the public API.

**Reply by elalish** [reply]
Yeah, I'm curious if this is an API that would be useful to add. Perhaps something that returns a vector of winding numbers matching an input vector of positions?

**Reply by pyenthu** [reply]
This would be a good idea. This is what I am trying to achieve. 
1. Get an array of positions from the object.
2. From another set of unique points, determine if those points lie inside the manifold or not.
3. Create an index of points to specifically "point" and vertices that I can use to morph.
4. I want to this in the gpu. So I need an array of returned checked function. 
But there may be other uses.

On another note, is there anyway we can use the glMatrix or math.gl library with the meshes for warping? Or modify the array in the gpu.js functions?

**Reply by elalish** [reply]
Are you using our JS bindings? Have you tried our `warp()` function? When you determine if your points are inside or outside the manifold, what do you do with that information? Do you only morph verts that are inside the manifold?

We use glMatrix in ManifoldCAD.org. I haven't heard of math.gl or gpu.js.

**Reply by pyenthu** [reply]
I am using the Warp Function. 
What I want to be able to do is to use the warp function via the gpu with parallel threads. This possible via a library called the gpu.js or using math.gl. 
Right now it looks like the warp function is sending a call for each vertex serially to the warp function with a lot of overhead. 
What I am looking to do is to flatten the vertices and send to the gpu in a single structured array so we can exploit the math from the hour itself.
Did I explain myself?
In threejs I could do this by taking the positions and sending them to the guy then just assigning the morphed vertices back to the mesh. 
Would it be possible for me to replace all the verts in a single line of assignment.

**Reply by pyenthu** [reply]
https://www.npmjs.com/package/math.gl?activeTab=readme

**Reply by pyenthu** [reply]
https://gpu.rocks/

**Reply by pca006132** [reply]
It may be possible to do that, but it is not our priority for now because `warp` is rarely the performance bottleneck, and performance with GPU is actually not that simple. This library had CUDA backend previously, but was removed because we can get faster performance by better CPU parallelization. We will enable CPU parallelization on js when tbb support with Emscripten is more mature.

**Reply by elalish** [reply]
Actually, `WarpBatch()` was added for exactly this reason, but it looks like it's only been hooked up to the Python bindings so far - would you like to make a PR to add it to the WASM bindings too? Our WASM doesn't touch the GPU, but your batch function could, I believe. I'm also curious if we might need to think through some async issues...

**Reply by pyenthu** [reply]
Yes absolutely I can use the gpu.js function to modify this in async. See in three mesh one can run the array of positions to the gpu and it is really fast. then you get that as a batch and just assign the vertices, update the normals and it works fine. If you have the capability to access the vertices and modify them in a batch it would be fantastic. 
But the key is to be able to do this with an array of vertices in a parallel process on the gpu. That coupled with all the csg functionality you have would be a game changer.
The other option is that I build the object in three and use the "batch" process but then I lose the capability to do fast csg your library does. I want rhe functionality of the three morphing along with manifold csg. That is a very good combination.

**Reply by elalish** [reply]
SG - would you like to try exposing our `WarpBatch()` in our WASM? I *think* it shouldn't be hard - you can follow the existing `Warp()` API. You're ideal since you can test it and let us know if there are practical problems we haven't thought through.

**Reply by pyenthu** [reply]
Let me try. I am not that savvy with several technologies. If there is a guide to doing this let me know k will try.
